src/freertos_drivers/esp32/Esp32WiFiConfiguration.hxx

```diff
@@ -0,0 +1,139 @@
+
+    /// Visible description for the Hub Configuration group.
+    static constexpr const char *HUB_DESC = 
+        "Configuration settings for an OpenLCB Hub";
+
+    /// Visible name for the hub enable field.
+    static constexpr const char *HUB_ENABLE_NAME = "Enable Hub Mode";
+
+    /// Visible description for the hub enable field.
+    static constexpr const char *HUB_ENABLE_DESC = 
+        "Defines this node as a hub which can accept connections";
+
+    /// Visible name for the hub_listener_port field.
+    static constexpr const char *HUB_LISTENER_PORT_NAME =
+        "Hub Listener Port";
+
+    /// Visible name for the hub_listener_port field.
+    static constexpr const char *HUB_LISTENER_PORT_DESC = 
+        "Defines the TCP/IP listener port this node will use when operating "
+        "as a hub. Most of the time this does not need to be changed.";
+
+    /// Visible name for the link_config group.
+    static constexpr const char *UPLINK_NAME = "Node Uplink Configuration";
+
+    /// Visible name for the link_config group.
+    static constexpr const char *UPLINK_DESC =
+        "Configures how this node will connect to other nodes.";
+};
+
+/// CDI Configuration for an @ref Esp32WiFiManager managed hub.
+CDI_GROUP(HubConfiguration);
+/// Allows the node to become a Grid Connect Hub.
+CDI_GROUP_ENTRY(enable, openlcb::Uint8ConfigEntry,
+    Name(Esp32WiFiConfigurationParams::HUB_ENABLE_NAME),
+    Description(Esp32WiFiConfigurationParams::HUB_ENABLE_DESC),
+    Min(0), Max(1), Default(0),
+    MapValues(Esp32WiFiConfigurationParams::BOOLEAN_MAP));
+/// Specifies the port which should be used by the hub.
+CDI_GROUP_ENTRY(port, openlcb::Uint16ConfigEntry,
+    Name(Esp32WiFiConfigurationParams::HUB_LISTENER_PORT_NAME),
+    Description(Esp32WiFiConfigurationParams::HUB_LISTENER_PORT_DESC),
+    Min(1), Max(65535),
+    Default(openlcb::TcpClientDefaultParams::DEFAULT_PORT))
+/// Specifies the mDNS service name to advertise for the hub.
+CDI_GROUP_ENTRY(service_name, openlcb::StringConfigEntry<48>,
+    Name(openlcb::TcpClientDefaultParams::SERVICE_NAME),
+    Description(openlcb::TcpClientDefaultParams::SERVICE_DESCR));
+CDI_GROUP_END();
+
+/// CDI Configuration for an @ref Esp32WiFiManager managed node.
+CDI_GROUP(WiFiConfiguration);
+/// Allows the WiFi system to use power-saving techniques to conserve power
+/// when the node is powered via battery.
+CDI_GROUP_ENTRY(sleep, openlcb::Uint8ConfigEntry,
+    Name(Esp32WiFiConfigurationParams::WIFI_POWER_SAVE_NAME),
+    Description(Esp32WiFiConfigurationParams::WIFI_POWER_SAVE_DESC),
+    Min(0), Max(1), Default(0),
+    MapValues(Esp32WiFiConfigurationParams::BOOLEAN_MAP));
+/// CDI Configuration to enable this node to be a hub.
+CDI_GROUP_ENTRY(hub, HubConfiguration,
+    Name(Esp32WiFiConfigurationParams::HUB_NAME),
+    Description(Esp32WiFiConfigurationParams::HUB_DESC));
+/// CDI Configuration for this node's connection to an uplink hub.
+CDI_GROUP_ENTRY(uplink,
```

Comment:
Is there an enable/disable setting for the uplink?
Or should there be a single 3/4-way setting for wifi instead:
- disabled
- client
- server
- client + server

Reply:
Right now there are two types of nodes configurable with this CDI:
1) hub + uplink
2) uplink only

I think it would make sense to add the third option of:
3) hub only

Right now I'm just exposing uplink as TcpClientConfig, would it make sense to add an "enable" option to that config or should I add "uplink_enabled" here? Or should there be a "mode" entry that uses a map (uplink only, hub only, hub+uplink)?

Reply:
What do you envision the usecase for "disabled" to be?

Reply:
CAN-bus only.

Reply:
Ahh, that does make sense. A can node that might make use of TCP/IP. I don't think that will be a common use case but it can be supported. Let me move hub enabled up to the top level as an enum like entry containing: uplink enabled, hub enabled, uplink and hub enabled, disabled. That should cover the options above and when we move to the WiFiInterface model we could possibly add a third mode of operation for this code to use a wifi profile instead of passing SSID/password.